README.md
# Track-Requests [![NPM Version][npm-version-image]][npm-url] [![Build Status](https://travis-ci.com/omargho/track-requests.svg?branch=master)](https://travis-ci.com/omargho/track-requests) [![Coverage Status](https://coveralls.io/repos/github/omargho/track-requests/badge.svg?branch=master)](https://coveralls.io/github/omargho/track-requests?branch=master)

HTTP request tracker middleware for node.js

A middleware that tracks http requests and save them 
into [ElasticSearch](https://www.elastic.co/fr/products/elastic-stack) in order to use [Kibana](https://www.elastic.co/fr/products/kibana) as a
Dashboard to give you the needed analytics of your application (sign up numbers,
number of active users, errors tracking...)  
  
  
## Getting Started
Install the module with: `npm install track-requests`
<!-- eslint-disable no-unused-vars -->

```js
const express = require('express');
const app = express();
const trackRequests = require('track-requests');
const options = {
  elasticSearchOptions: {
    elasticSearchUrl: 'http://127.0.0.1:9200/'
  }
};
app.use(trackRequests(options));
```

## Documentation
### `trackRequests(options)`
Create a new track request middleware function using the given `options`.  
`options` is an object container for parameters used to customise the trackRequest middleware
 

### `options.getters`
`Object`- Contains functions used to generate the tracking document

#### `options.getters.getUser`
`Function`- return `String` the http req user's Id. it takes as parameter `req` (The HTTP request) and `res` (The HTTP response)

`default: () => 'anonymous'`
 
Usage example: 
```js
let options = {
  getters: {
    getUser: (req, res) => {
      return req.user.id;
    }
  }
};
```

#### `options.getters.getIp`
`Function`- return `String` the http req user's Ip. it takes as parameter `req` (The HTTP request) and `res` (The HTTP response)

`default: (req) => req.headers['x-real-ip']`

#### `options.getters.getUserAgent`
`Function`- return `String` the http req user's userAgent. it takes as parameter `req` (The HTTP request) and `res` (The HTTP response)

`default: (req) => req.headers['user-agent']`

#### `options.getters.getErrorMessage`
`Function`- return `String` the error message when an error happens on the controller. it takes as parameter `req` (The HTTP request) and `res` (The HTTP response)

`default: (req) => req.error`

#### `options.getters.getBody`
`Function`- return `Object` http request body. it takes as parameter `req` (The HTTP request) and `res` (The HTTP response)

`default: (req) => req.body`

#### `options.getters.getExtraFields`
`Function`- return `Object` extra fields to be added to the generated tracking document. it takes as parameter `req` (The HTTP request) and `res` (The HTTP response)  
`PS: make sure to add a custom options.elasticSearchOptions.indexProperties that contains the extra fields parmeters`

Usage example: 
```js
let options = {
  elasticSearchOptions: {
    indexProperties: {
      url: {'type': 'keyword'},
      method: {'type': 'keyword'},
      user: {'type': 'keyword'},
      ip: {'type': 'ip'},
      statusCode: {'type': 'short'},
      processingTime: {'type': 'integer'},
      createdAt: {'type': 'date'},
      error: {'type': 'text'},
      userAgent: {'type': 'text'},
      body: {'type': 'nested'},
      myCustomField: {'type': 'text'}
    }
  },
  getters: {
    getExtraFields: (req, res) => {
      return {
        myCustomField:' myCustomFieldValue'
      }
    }
  }
};
```

### `options.routesToIgnore`
`Array`- array of `String`|`Regex`|`Object` of the endpoints that you want the tracker to ignore them

- `String` it should match the exact endpoint's url to ignore it
- `Regex`  it should verify the endpoint's url to ignore it
- `Object` all the parameters should matches to ignore the endpoint
    - url `String` it should match the exact endpoint
    - regex `Regex` it should verify the endpoint
    - statusCode `Number` it should match the HTTP response's Status Code
    - method `String` it should match the HTTP request's method

Usage example: 
```js
let options = {
  routesToIgnore:[
    /user/, //ignore all the endpoint that verify this regex 
    '/organization/member',//ignore this exact endpoint  
    {url: '/healthcheck', method: 'GET'},// ignore the GET requests to this exact endpoint    
    {regex: /search/, method: 'POST', statusCode: 404}// ignore the POST with a 404 response status code that verify this regex  
  ]
};
``` 

### `options.elasticSearchOptions`
`Object`- Contains the parameters used to customise the elasticSearch client

#### `options.elasticSearchOptions.elasticSearchUrl`
`String`- ElasticSearch's Url  `(default: http://127.0.0.1:9200/)`

#### `options.elasticSearchOptions.indexName`
`String`- ElasticSearch's Index Name `(default: tracking)`

#### `options.elasticSearchOptions.indexProperties`
`Object`- ElasticSearch's Index Properties, it represent the schema of the object that will be saved into elasticSearch Document.  
 It's used to:
  - setup the elasticSearch index when it does not already exists   
  - make sure that the generated document that will be inserted into ES have the specified schema (any other parameter of the generated document will be deleted before the insertion)

`default:`   
```json
{
  "url": {"type" : "keyword"},
  "method": {"type" : "keyword"},
  "user": {"type": "keyword"},
  "ip": {"type": "ip"},
  "statusCode": {"type": "short"},
  "processingTime": {"type": "integer"},
  "createdAt": {"type": "date"},
  "error": {"type": "text"},
  "userAgent": {"type": "text"},
  "body": {"type": "nested"}
}  
``` 

[MIT](LICENSE)

[npm-url]: https://www.npmjs.com/package/track-requests
[npm-version-image]: https://badgen.net/npm/v/track-requests


HISTORY.md
unreleased
==========

  * Support koa
  
1.0.2 / 2020-02-2
==================
  * Improve docs
  * Add Tests + Code Coverage + CI 
  * Add ignore routes Feature
    
1.0.1 / 2020-01-25
==================
  * Fixes
    
1.0.0 / 2020-01-25
==================

  * Initial release


test/basicTests.js
'use strict';

const serverFactory = new (require('./utils/serverFactory'))();
const trackerRequests = require('..');
const request = require('request');
const assert = require('assert');

describe('Basic tests', () => {
  const options = {
    elasticSearchOptions: {
      elasticSearchUrl: 'http://127.0.0.1:9200/',
      indexName: 'test-tracking'
    }
  };

  before(done => {
    let {elasticSearchUrl, indexName} = options.elasticSearchOptions;
    //delete ES data
    request.delete(elasticSearchUrl + indexName, () => {
      done();
    });
  });

  after(done => {
    let {elasticSearchUrl, indexName} = options.elasticSearchOptions;
    //delete ES data
    request.delete(elasticSearchUrl + indexName, () => {
      setTimeout(done, 300);
    });
  });

  serverFactory.createSever(3000, [trackerRequests(options)],
    [
      {method: 'get', endpoint: '/', controller: (req, res) => res.send('Hakuna Matata')}
    ]);

  it('check if server started correctly', (done) => {
    request.get('http://127.0.0.1:3000/', (err, response, body) => {
      assert.equal(response.statusCode, 200);
      assert.equal(body, 'Hakuna Matata');
      done();
    });
  });

});

test/trackerTests.js
'use strict';

const Tracker = require('../lib/tracker');
const assert = require('assert');

describe('tracker basic Tests', () => {

  const options = {
    elasticSearchOptions: {
      elasticSearchUrl: 'http://127.0.0.1:9200/',
      indexName: 'test-tracking'
    },
    routesToIgnore: [/user/, '/organization/member', {url: '/healthcheck', method: 'GET'}, {regex: /search/, method: 'POST', statusCode: 404}]
  };
  const tracker = new Tracker(options);

  it('generateTrackingDoc GET request example', (done) => {
    let req = {url: '/test', method: 'GET', headers: {'user-agent': 'scraper', 'x-real-ip': '127.0.0.1'}};
    let res = {statusCode: 200};
    let startTime = Date.now();
    let doc = tracker.generateTrackingDoc(req, res, startTime);
    let expectedDoc = {
      url: req.url,
      method: req.method,
      user: 'anonymous',
      ip: req.headers['x-real-ip'],
      statusCode: res.statusCode,
      processingTime: 0,
      createdAt: startTime,
      userAgent: req.headers['user-agent'],
      error: null,
      body: null
    };
    assert(expectedDoc.processingTime < 5);
    expectedDoc.createdAt = doc.createdAt;
    expectedDoc.processingTime = doc.processingTime;
    assert.deepEqual(expectedDoc, doc);

    done();
  });

  it('generateTrackingDoc POST request example', (done) => {
    let req = {url: '/test', method: 'POST', body: {name: 'fartas'}, headers: {'user-agent': 'scraper', 'x-real-ip': '127.0.0.1'}};
    let res = {statusCode: 200};
    let startTime = Date.now();
    let doc = tracker.generateTrackingDoc(req, res, startTime);
    let expectedDoc = {
      url: req.url,
      method: req.method,
      user: 'anonymous',
      ip: req.headers['x-real-ip'],
      statusCode: res.statusCode,
      processingTime: 0,
      createdAt: startTime,
      userAgent: req.headers['user-agent'],
      body: req.body,
      error: null
    };
    assert(expectedDoc.processingTime < 5);
    expectedDoc.createdAt = doc.createdAt;
    expectedDoc.processingTime = doc.processingTime;
    assert.deepEqual(expectedDoc, doc);

    done();
  });

  it('generateTrackingDoc err request example', (done) => {
    let req = {url: '/test', method: 'POST', body: {name: 'fartas'}, headers: {'user-agent': 'scraper', 'x-real-ip': '127.0.0.1'}, error: 'internal error'};
    let res = {statusCode: 500};
    let startTime = Date.now();
    let doc = tracker.generateTrackingDoc(req, res, startTime);
    let expectedDoc = {
      url: req.url,
      method: req.method,
      user: 'anonymous',
      ip: req.headers['x-real-ip'],
      statusCode: res.statusCode,
      processingTime: 0,
      createdAt: startTime,
      userAgent: req.headers['user-agent'],
      body: req.body,
      error: req.error
    };
    assert(expectedDoc.processingTime < 5);
    expectedDoc.createdAt = doc.createdAt;
    expectedDoc.processingTime = doc.processingTime;
    assert.deepEqual(expectedDoc, doc);

    done();
  });

  it('shouldIgnoreRoutes should match regex', (done) => {
    let req = {url: '/test/user', method: 'POST', headers: {}};
    let res = {statusCode: 201};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(result);
    done();
  });

  it('shouldIgnoreRoutes should not match', (done) => {
    let req = {url: '/organization/members', method: 'POST', headers: {}};
    let res = {statusCode: 201};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(!result);
    done();
  });

  it('shouldIgnoreRoutes should match string', (done) => {
    let req = {url: '/organization/member', method: 'POST', headers: {}};
    let res = {statusCode: 201};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(result);
    done();
  });

  it('shouldIgnoreRoutes should not match object', (done) => {
    let req = {url: '/healthcheck', method: 'POST', headers: {}};
    let res = {statusCode: 201};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(!result);
    done();
  });

  it('shouldIgnoreRoutes should match object', (done) => {
    let req = {url: '/healthcheck', method: 'GET', headers: {}};
    let res = {statusCode: 201};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(result);
    done();
  });

  it('shouldIgnoreRoutes should not match object with regex', (done) => {
    let req = {url: '/search', method: 'POST', headers: {}};
    let res = {statusCode: 401};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(!result);
    done();
  });

  it('shouldIgnoreRoutes should match object with regex', (done) => {
    let req = {url: '/search', method: 'POST', headers: {}};
    let res = {statusCode: 404};
    let result = tracker.shouldIgnoreRoutes(req, res);
    assert(result);
    done();
  });

});

describe('tracker custom Getters Tests', () => {

  const options = {
    elasticSearchOptions: {
      elasticSearchUrl: 'http://127.0.0.1:9200/',
      indexName: 'test-tracking'
    },
    getters: {
      getUser: (req) => req.user && req.user.id || 'sankou7',
      getIp: (req) => req.ip,
      getErrorMessage: (req, res) => res.error,
      getExtraFields: (req) => {
        return {
          params: req.params,
          query: req.query,
          hello: 'hello ' + (req.user && req.user.id || 'sankou7')
        };
      }
    }
  };

  const tracker = new Tracker(options);

  it('generateTrackingDoc GET request example', (done) => {
    let req = {url: '/test', method: 'GET', ip: '8.8.8.8', user: {id: 'balbaz'}, headers: {'user-agent': 'scraper', 'x-real-ip': '127.0.0.1'}};
    let res = {statusCode: 200};
    let startTime = Date.now();
    let doc = tracker.generateTrackingDoc(req, res, startTime);
    let expectedDoc = {
      url: req.url,
      method: req.method,
      user: 'balbaz',
      ip: req.ip,
      statusCode: res.statusCode,
      processingTime: 0,
      createdAt: startTime,
      userAgent: req.headers['user-agent'],
      error: undefined,
      body: undefined,
      query: undefined,
      params: undefined,
      hello: 'hello balbaz'
    };
    assert(expectedDoc.processingTime < 5);
    expectedDoc.createdAt = doc.createdAt;
    expectedDoc.processingTime = doc.processingTime;
    assert.deepEqual(expectedDoc, doc);

    done();
  });

  it('generateTrackingDoc POST request example', (done) => {
    let req = {
      url: '/test', method: 'GET', ip: '8.8.8.8', headers: {'user-agent': 'scraper', 'x-real-ip': '127.0.0.1'},
      body: {ok: 'ok'},
      params: {ok: 'not ok'},
      query: {id: 1}
    };
    let res = {statusCode: 200};
    let startTime = Date.now();
    let doc = tracker.generateTrackingDoc(req, res, startTime);
    let expectedDoc = {
      url: req.url,
      method: req.method,
      user: 'sankou7',
      ip: req.ip,
      statusCode: res.statusCode,
      processingTime: 0,
      createdAt: startTime,
      userAgent: req.headers['user-agent'],
      error: undefined,
      body: req.body,
      query: req.query,
      params: req.params,
      hello: 'hello sankou7'
    };
    assert(expectedDoc.processingTime < 5);
    expectedDoc.createdAt = doc.createdAt;
    expectedDoc.processingTime = doc.processingTime;
    assert.deepEqual(expectedDoc, doc);

    done();
  });

  it('generateTrackingDoc err request example', (done) => {
    let req = {
      url: '/test', method: 'GET', ip: '8.8.8.8', headers: {'user-agent': 'scraper', 'x-real-ip': '127.0.0.1'},
      body: {ok: 'ok'},
      params: {ok: 'not ok'},
      query: {id: 1}
    };
    let res = {statusCode: 404, error: 'balbaz not found'};
    let startTime = Date.now();
    let doc = tracker.generateTrackingDoc(req, res, startTime);
    let expectedDoc = {
      url: req.url,
      method: req.method,
      user: 'sankou7',
      ip: req.ip,
      statusCode: res.statusCode,
      processingTime: 0,
      createdAt: startTime,
      userAgent: req.headers['user-agent'],
      error: 'balbaz not found',
      body: req.body,
      query: req.query,
      params: req.params,
      hello: 'hello sankou7'
    };
    assert(expectedDoc.processingTime < 5);
    expectedDoc.createdAt = doc.createdAt;
    expectedDoc.processingTime = doc.processingTime;
    assert.deepEqual(expectedDoc, doc);

    done();
  });
});

test/esClientTests.js
'use strict';

const ElasticSearchClient = require('../lib/elasticSearchClient');
const request = require('request');
const assert = require('assert');

describe('esClient basic Tests', () => {
  const options = {
    elasticSearchUrl: 'http://127.0.0.1:9200/',
    indexName: 'test-tracking'
  };
  const esClient = new ElasticSearchClient(options);

  before(done => {
    //delete ES data
    request.delete(options.elasticSearchUrl + options.indexName, () => {
      done();
    });
  });

  after(done => {
    let {elasticSearchUrl, indexName} = options;
    //delete ES data
    request.delete(elasticSearchUrl + indexName, () => {
      done();
    });
  });

  it('check esClient init', (done) => {
    assert.equal(esClient.indexName, options.indexName);
    assert.equal(esClient.esUrl, options.elasticSearchUrl);
    assert(esClient.indexProperties);
    done();
  });

  it('healthCheck', (done) => {
    esClient.healthCheck().then(body => {
      assert.equal(body.tagline, 'You Know, for Search');
      done();
    }).catch(done);
  });

  it('setMapping', (done) => {
    esClient.setMapping().then(done).catch(done);
  });

  it('getMapping', (done) => {
    esClient.getMapping().then(body => {
      assert.deepEqual(esClient.indexProperties, body);
      done();
    }).catch(done);
  });

  it('getMapping when index not found(should execute setMapping)', (done) => {
    //delete ES data
    request.delete(options.elasticSearchUrl + options.indexName, () => {
      esClient.getMapping().then(body => {
        assert.deepEqual(esClient.indexProperties, body);
        done();
      }).catch(done);
    });
  });

  it('insertDocument', (done) => {
    let document = {
      url: '/host',
      method: 'GET',
      user: 'sankou7',
      ip: '127.0.0.1',
      statusCode: 200,
      processingTime: 10,
      createdAt: Date.now(),
      userAgent: 'scraper'
    };
    esClient.insertDocument(document).then(doc => {
      assert.notDeepStrictEqual(doc, document);
      done();
    }).catch(done);
  });

});

describe('esClient Bad esUrl Tests', () => {
  const options = {
    elasticSearchUrl: 'http://127.0.0.1:9201/',
    indexName: 'test-tracking'
  };
  const esClient = new ElasticSearchClient(options);

  it('healthCheck should throw an error', (done) => {
    esClient.healthCheck().then(body => {
      done(body);
    }).catch(err => {
      assert(err.code, 'ECONNREFUSED');
      done();
    });
  });

  it('setMapping should throw an error', (done) => {
    esClient.setMapping().then(done).catch(err => {
      assert(err.message.toString().includes('Track-Requests: error while mappings'));
      done();
    });
  });

  it('getMapping should throw an error', (done) => {
    esClient.getMapping().then(done).catch(err => {
      assert(err.message.toString().includes('Track-Requests: error while fetching'));
      done();
    });
  });

  it('insertDocument should return an error', (done) => {
    esClient.insertDocument({url: 'test'}).then(err => {
      assert.equal(err.error, true);
      done();
    });
  });

});

describe('esClient custom index properties', () => {
  const options = {
    elasticSearchUrl: 'http://127.0.0.1:9200/',
    indexName: 'test-tracking',
    indexProperties: {
      url: {'type': 'keyword'},
      method: {'type': 'keyword'},
      user: {'type': 'keyword'},
      params: {'type': 'nested'}
    }
  };
  before(done => {
    //delete ES data
    request.delete(options.elasticSearchUrl + options.indexName, () => {
      done();
    });
  });

  after(done => {
    let {elasticSearchUrl, indexName} = options;
    //delete ES data
    request.delete(elasticSearchUrl + indexName, () => {
      done();
    });
  });

  const esClient = new ElasticSearchClient(options);

  it('check esClient indexProperties', (done) => {
    assert.deepEqual(esClient.indexProperties, options.indexProperties);
    done();
  });

  it('healthCheck', (done) => {
    esClient.healthCheck().then(body => {
      assert.equal(body.tagline, 'You Know, for Search');
      done();
    }).catch(done);
  });

  it('setMapping', (done) => {
    esClient.setMapping().then(done).catch(done);
  });

  it('insertDocument', (done) => {
    let document = {
      url: '/host',
      method: 'GET',
      user: 'sankou7',
      ip: '127.0.0.1',
      statusCode: 200,
      processingTime: 10,
      createdAt: Date.now(),
      userAgent: 'scraper',
      params: {data: 'test'}
    };
    let expectedDoc = {};
    Object.keys(options.indexProperties).forEach(key => {
      expectedDoc[key] = document[key];
    });
    esClient.insertDocument(document).then(doc => {
      assert.deepEqual(expectedDoc, doc);
      done();
    }).catch(done);
  });

});

lib/tracker.js
'use strict';
const ElasticSearchClient = require('./elasticSearchClient');

class Tracker {

  constructor(opts = {getters: {}}) {
    this.active = true;
    this.logger = opts.logger || console.error;
    this.routesToIgnore = opts.routesToIgnore || [];
    this.getters = {
      getUrl: (req) => req.url,
      getMethod: (req) => req.method,
      getUser: () => 'anonymous',
      getIp: (req) => req.headers['x-real-ip'],
      getStatusCode: (req, res) => res.statusCode === 304 ? 200 : res.statusCode,
      getProcessingTime: (startTime) => Date.now() - startTime,
      getCreatedAt: Date.now,
      getErrorMessage: (req) => req.error,
      getUserAgent: (req) => req.headers['user-agent'],
      getBody: (req) => req.body,
      ...opts.getters
    };

    this.esClient = opts.dbClient || new ElasticSearchClient({...opts.elasticSearchOptions, logger: this.logger});

    this.generateTrackingDoc = opts.generateTrackingDoc || (
      (req, res, startTime) => {
        try {
          let trackingDoc = {
            url: this.getters.getUrl(req, res),
            method: this.getters.getMethod(req, res),
            user: this.getters.getUser(req, res),
            ip: this.getters.getIp(req, res),
            statusCode: this.getters.getStatusCode(req, res),
            processingTime: this.getters.getProcessingTime(startTime),
            createdAt: this.getters.getCreatedAt(req, res),
            error: this.getters.getErrorMessage(req, res),
            userAgent: this.getters.getUserAgent(req, res),
            body: this.getters.getBody(req, res),
            ...opts.getters && opts.getters.getExtraFields && opts.getters.getExtraFields(req, res)
          };

          return trackingDoc;
        } catch (err) {
          this.logger('track-requests: cannot generate tracking doc', err);
        }
      }
    );

    this.shouldIgnoreRoutes = (req, res) => {
      let url = this.getters.getUrl(req, res);
      return this.routesToIgnore.some((filter) => {
        if (filter instanceof RegExp)
          return filter.test(url);
        if (typeof filter == 'object')
          return (!filter.url || filter.url == url) && (!filter.regex || filter.regex.test(url)) && (!filter.method || filter.method == this.getters.getMethod(req))
            && (!filter.statusCode || filter.statusCode && filter.statusCode == this.getters.getStatusCode(req, res));

        return filter == url;

      });
    };
  }
}

module.exports = Tracker;

lib/elasticSearchClient.js
'use strict';

const request = require('request');

const requestPromise = (options) => {
  return new Promise((resolve, reject) => {
    request(options, (err, res, body) => {
      if (err) return reject(err);
      if (res.statusCode >= 400) return reject({body, status: res.statusCode, headers: res.headers});
      return resolve({body, status: res.statusCode, headers: res.headers});
    });
  });
};

class ElasticSearchClient {
  constructor(opts = {}) {
    this.logger = opts.logger || (() => {});
    this.esUrl = opts.elasticSearchUrl || 'http://127.0.0.1:9200/';
    this.indexName = opts.indexName || 'tracking';
    this._esVersion = '7.5.0';
    this.indexProperties = opts.indexProperties || {
      url: {'type': 'keyword'},
      method: {'type': 'keyword'},
      user: {'type': 'keyword'},
      ip: {'type': 'ip'},
      statusCode: {'type': 'short'},
      processingTime: {'type': 'integer'},
      createdAt: {'type': 'date'},
      error: {'type': 'text'},
      userAgent: {'type': 'text'},
      body: {'type': 'nested'}
    };
  }

  async getMapping() {
    try {
      let {body: mapping} = await requestPromise({method: 'GET', url: this.esUrl + this.indexName + '/_mapping', json: true});
      let mappings = mapping[this.indexName].mappings;
      return this._esVersion < '6.0.0' ? mappings.default.properties : mappings.properties;
    } catch (e) {
      if (e.status === 404 && e.body.error && e.body.error.type == 'index_not_found_exception') {
        await this.setMapping();
        return await this.getMapping();
      }
      throw Error('Track-Requests: error while fetching mapping:' + (e.body && JSON.stringify(e.body) || e));
    }
  }

  async healthCheck() {
    let {body} = await requestPromise({method: 'GET', url: this.esUrl, json: true});
    this._esVersion = body.version.number;
    return body;
  }

  async setMapping() {
    try {
      await requestPromise({
        method: 'PUT',
        url: this.esUrl + this.indexName,
        body: {
          mappings: this._esVersion < '6.0.0'
            ? {default: {properties: this.indexProperties}}
            : {properties: this.indexProperties}
        },
        json: true
      });
    } catch (e) {
      throw Error('Track-Requests: error while mappings initializing :' + (e.body && JSON.stringify(e.body) || e));
    }
  }

  async insertDocument(document) {
    try {
      let cleanedDoc = {};
      Object.keys(this.indexProperties).forEach(key => {
        cleanedDoc[key] = document[key] || null;
      });
      await requestPromise({
        method: 'POST',
        url: this.esUrl + this.indexName + (this._esVersion < '6.0.0' ? '/default' : '/_doc'),
        body: cleanedDoc,
        json: true
      });

      return cleanedDoc;
    } catch (e) {
      this.logger('Track-Requests: error while inserting a document :' + (e.body && JSON.stringify(e.body) || e));
      return {error: true, message: 'Track-Requests: error while inserting a document :' + (e.body && JSON.stringify(e.body) || e)};
    }
  }
}

module.exports = ElasticSearchClient;